### ISSUE TITLE
Service with http call in constructor is injected as undefined in HttpInterceptor

### ISSUE THREAD

**Issue by JGH153** [angular/angular#25590]
## I'm submitting a...

<pre><code>
[ ] Regression (a behavior that used to work and stopped working in a new release)
[x] Bug report  
[ ] Performance issue
[ ] Feature request
[ ] Documentation issue or request
[ ] Support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
[ ] Other... Please describe:
</code></pre>

## Current behavior
I have a service which executes an http get request in its constructor. When the service is injected into a constructor of a Component it's working, but when it's injected into the constructor of a HttpInterceptor it's undefined.

## Expected behavior
That the instance of the service is not undefined when it's injected into the HttpInterceptor constructor.

## Minimal reproduction of the problem with instructions
Open this example which highlights the problem in the console: https://stackblitz.com/edit/angular-cbhw93

## What is the motivation / use case for changing the behavior?
I'm assuming this is a bug, since I haven't found any documentation stating that preforming an http request in the constructor of a Service isn't supported. 

## Environment
<pre><code>
Angular version: 6.1.3

Browser:
- [x] Chrome (desktop) version 68
- [ ] Chrome (Android) version XX
- [ ] Chrome (iOS) version XX
- [ ] Firefox version XX
- [ ] Safari (desktop) version XX
- [ ] Safari (iOS) version XX
- [ ] IE version XX
- [ ] Edge version XX
 
For Tooling issues:
- Node version: 8.9.1
- Platform:  Windows
</code></pre>

Others:
The bug is reproducible on stackblitz using the latest version of angular

**Comment by pidoubleyou** [comment]
You have a problem with the creation order of the service and the interceptor.
The service creation has to be finished before the interceptor is created.

If you call the http request in the service constructor, the interceptor instance is created during the service creation.

Extract the http request in a method and make the call in the component.

**Comment by JGH153** [comment]
The constructor in the Service only initiates an async task that does not need to be completed for the the service to be ready. It would be really great to not have to move that http call to a components OnInit, or at least have the choice as it works for all cases except when used in an HttpInterceptor. 

Wrapping the HTTP call in the Service in a setTimeout also solves the problem, but would still prefer to avoid any workarounds.

**Comment by trotyl** [comment]
Duplicate of https://github.com/angular/angular/issues/24306 (The title is misleading, but same circular dependency problem).

**Comment by JGH153** [comment]
The behavior changes if I switch back to the old way of providing the services (in the module, as opposed to providedIn: 'root'). The old way causes an infinite loop where a new instance of the service is created each time the interceptor intercepts a new request, and each new instance of the service fires a new http request. It never stabilises before exceeding max call stack size. 

Stackblitz example: https://stackblitz.com/edit/angular-s99pyd

**Comment by trotyl** [comment]
It will be error anyway as the usage is not valid, but the error message could indeed be improved.

**Comment by JGH153** [comment]
Making an HTTP request from a Service constructor is not supported? If that's the case, then the docs should specify that at some point, and maybe emitt a error message if the Service fail to initialize before its injected. It is also really inconsistent as it only fails to work when the Service is injected into a interceptor, and it can easily be solved by a setTimeout around the http call.

**Comment by trotyl** [comment]
Making circular dependency in DI is not supported. Nothing to do with HTTP.

**Comment by alex88** [comment]
Same here, how can have a service (in this case an authentication service) make an http request (in this case, fetch current user) while having an http interceptor (in this case one that calls auth service logout if it intercepts a 401)?
In my case using a `setTimeout(.., 100)` is enough, but is there a better solution?
For example in a component I would do this in `ngOnInit`

**Comment by JGH153** [comment]
As long as this bug is still present, and we don't have ngOnInit in services I see two solutions:
1. Use a setTimeout in the service constructor (you can reduce it to setTimeout(.., 0))
2. Have a components ngOnInit call the services function for fetching the user
For handling a 401 and logging the user out you can catch the 401in the interceptor using .pipe(catchError(...))

**Comment by rekoch** [comment]
Don't know if it helps to another one. But I considered the problem while using 2 services inside a Interceptor. After a debugging and trial and error session I figured out, that the constructor is used in well ordner. So before the function "intercept()" it uses the constructor. At the constructor time, both services were injected right and I got an instance. But while the intercept function, they where undefined. So I tried another way. Instead of having two local variables for the services I just used static variables. In the constructor I referenced the injected services to my static variables. After that I could get the right instance of the service.

So here is a code snippet for that solution (here the handleError is static too because an instance does not make sense any more):

`@Injectable()
export class ErrorInterceptor implements HttpInterceptor {
  static translateService;

  constructor(translateService: TranslateService) {
    ErrorInterceptor.translateService = translateService;
  }

  static handleError(error: HttpErrorResponse) {
    if (error.error instanceof ErrorEvent) {
      // A client-side or network error occurred. Handle it accordingly.
      console.error('An error occurred:', error.error.message);
    } else {
      // The backend returned an unsuccessful response code.
      // The response body may contain clues as to what went wrong,
      const titleTranslated = ErrorInterceptor.translateService.instant('MESSAGES.ERROR_TITLE');
        console.log(titleTranslated)
      };
    }
    return throwError(
      titleTranslated + 'Something happened while calling remote call.');
  }

  intercept(req: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
    return next.handle(req)
      .pipe(
        catchError(ErrorInterceptor.handleError)
      );
  }
}
`

**Comment by vbourdeix** [comment]
My solution, even if it feels dirty to me, is to give access to the asynchronous data (a user token validated on the server in my case) in an observable and to use that observable inside the HttpInterceptor. This same use case is described by Daniel Crisp on this blog post : https://medium.com/@danielcrisp/async-http-interceptors-with-angular-4-3-9e6e795da562

**Comment by sasensi** [comment]
I am encountering the same problem: https://stackblitz.com/edit/interceptor-dependency
My use case is also to validate a stored token.
I am using the `setTimeout` workaround for now.

**Comment by rekoch** [comment]
@sasensi 
You're problem is more related to the fact, that you're calling the myService too early. I just moved your httpClient Call away from the constructor to a new function. Inside the functio you then call the API behind ( "https://yesno.wtf/api"). Then everything is working fine. Be carefull with calling things directly in the constructor. In Angular some part of the App are not yet there due to the boostrapping.

See here https://stackblitz.com/edit/interceptor-dependency-vrgjmy

Hope it helps

**Comment by sasensi** [comment]
@rekoch 
Yes, indeed, this is another possible workaround, thank you.

**Comment by trotyl** [comment]
> @rekoch
> Yes, indeed, this is another possible workaround, thank you.

That's the expected usage, not a workaround. One can never have eager circular dependencies.

**Comment by JGH153** [comment]
Getting support for ngOnInit in the service would help a lot, assuming ngOnInit is executed after the current service is done constructing, and are available for DI.

https://github.com/angular/angular/issues/23235

**Comment by amaust82** [comment]
FWIW, I bumped into this issue and found that adding the service as a dep seems to work for me.  So in your module, set up your provider like so:

```
@NgModule({
  providers: [
    AuthInterceptor,
    LoggingService,
    {
      provide: HTTP_INTERCEPTORS,
      useClass: AuthInterceptor,
      multi: true,
      deps: [LoggingService]
    }
  ]
  ...
})
```

Then, in my constructor I can import it.
```
@Injectable({
	providedIn: 'root',
})
export class AuthInterceptor implements HttpInterceptor {
  constructor(private logger: LoggingService) {}

  intercept(req: HttpRequest<any>, next: HttpHandler) {
    this.logger.trace('My output'); // This works for me now
  }
}
```

I haven't dug deep yet, but it is working, though if you have multiple deps, the order seems to matter. That is unconfirmed, though.

**Comment by petebacondarwin** [comment]
In Angular 12 you now get a clear error message due to the circular dependency.
See https://stackblitz.com/edit/angular-ivy-58vjwq?devtoolsheight=33&file=src/app/app.module.ts

Since circular dependencies are an invalid DI setup in Angular, there is a clear message now, and this issue did not get much community support as a feature, I am going to close it.
